src/hotspot/share/cds/heapShared.cpp

```diff
@@ -781,6 +781,137 @@ void HeapShared::copy_objects() {
   }
 }
 
+// Mark the classes of all objects that are scanned HeapShared::find_all_aot_initialized_classes()
+class HeapShared::AOTInitializedClassScanner : public BasicOopIterateClosure {
+  bool _made_progress;
+
+  template <class T> void check(T *p) {
+    oop obj = HeapAccess<>::oop_load(p);
+    if (!java_lang_Class::is_instance(obj)) {
+      // Don't scan the mirrors, as we may see an orig_mirror while scanning
+      // the object graph, .... TODO more info
+      _made_progress |= HeapShared::scan_for_aot_initialized_classes(obj);
+    }
+  }
+
+public:
+  AOTInitializedClassScanner() : _made_progress(false) {}
+  void do_oop(narrowOop *p) { check(p); }
+  void do_oop(      oop *p) { check(p); }
+  bool made_progress() { return _made_progress; }
+};
+
+static bool mark_for_aot_initialization(InstanceKlass* buffered_ik) {
+  if (buffered_ik->name()->equals("java/lang/String")) {
```

Comment:
Why is String not AOT initialized?  We preserve Strings (and the whole intern table) in the AOTCache.  Given we have many instances of String objects in our cache, why wouldn't we AOT init String?

Is String's initialization handled specially elsewhere?

Reply:
I was just being too conservation, as `String::COMPACT_STRINGS`, etc, depends on VM flags. I forgot that CDS already checks that the `CompactStrings` flag must be used between dump time and run time.

I've include String in the latest JEP 483 PR and found no issues.